src/assign.c

```diff
@@ -811,322 +797,201 @@ const char *memrecycle(SEXP target, SEXP where, int start, int len, SEXP source,
     // checks up front, otherwise we'd need checks twice in the two branches that cater for 'where' or not
-    bool pun = false;
+    // TODO:  verbose message and/or strict option for advanced users to ensure types match
+    //        only call getOption (small cost in finding the option value) at this point when there is a type mismatch
     switch(TYPEOF(target)) {
     case LGLSXP:
       if (isInteger(source)) {
         const int *sD = INTEGER(source);
         for (int i=0; i<slen; ++i) {
           const int val = sD[i];
-          if (val!=0 && val!=1 && val!=NA_INTEGER) error("Cannot assign %d to a logical column.", val);
+          if (val!=0 && val!=1 && val!=NA_INTEGER) {
+            warning("Non-zero %d taken as TRUE when assigning to logical column. Please use as.logical() to avoid this warning"
+                    " and to express the intent to help those reading this code in future.", val);
+            // was warning too in v1.12.2 but it was much longer and less helpful/specific
+            break; // just warn on the first
+          }
         }
-        pun = true;
       } else if (isReal(source)) {   // TODO: cater for integer64 here
         const double *sD = REAL(source);
         for (int i=0; i<slen; ++i) {
           const double d = sD[i];
-          if (!ISNAN(d) && d!=0.0 && d!=1.0) error("Cannot assign %f to a logical column.", d);
+          if (!ISNAN(d) && d!=0.0 && d!=1.0) {
+            warning("Non-zero %f taken as TRUE when assigning to a logical column. Please use as.logical() to avoid this warning"
+                    " and to express the intent to help those reading this code in future.", d);
+            break;
+          }
         }
-        pun = true;
       }
       break;
     case INTSXP:
       if (isReal(source)) {
         int w = INTEGER(isReallyReal(source))[0];  // first fraction present (1-based), 0 if none
         if (w>0) {
           warning("Coerced double RHS to integer to match the type of the target column (column %d named '%s'). One or more RHS values contain fractions which have been lost; e.g. item %d with value %f has been truncated to %d.", colnum, colname, w, REAL(source)[w-1], (int)REAL(source)[w-1]);
+          // same warning text as v1.12.2 so as to reduce diff in tests. TODO: shorten text in future to put truncated %f at the begnning of the message.
         }
-        pun = true;
-      } else if (isLogical(source)) {
-        pun = true;
       }
       break;
     case REALSXP:
-      pun = isLogical(source) || isInteger(source) || TYPEOF(source)==RAWSXP;
-      break;
-    case VECSXP:
-      pun = true;
+      if (targetIsI64 && isReal(source) && !sourceIsI64) {
+        int firstReal=0;
+        if ((firstReal=INTEGER(isReallyReal(source))[0])) {
+          sprintf(memrecycle_message, "coerced to integer64 but contains a non-integer value (%f at position %d); precision lost.", REAL(source)[firstReal-1], firstReal);
```

Comment:
shouldn't we use `snprintf` here?

Reply:
good catch on `snprintf`. done.

> ideally, for 1.13.0 we could try to take out coercion from assign and make it re-usable

that's what I did in this PR iiuc; i.e.  the assign code is taken out of assign() and placed into memrecycle().  memrecycle() is callable from elsewhere.